rs/p2p/artifact_manager/src/lib.rs

```diff
@@ -188,10 +203,20 @@ fn process_messages<Artifact: IdentifiableArtifact + 'static>(
         };
 
         let batched_artifact_events = current_thread_rt.block_on(async {
-            match timeout(recv_timeout, receiver.recv()).await {
+            let mut inbound_stream = std::pin::Pin::new(&mut inbound_stream);
+            match timeout(recv_timeout, inbound_stream.next()).await {
                 Ok(Some(artifact_event)) => {
                     let mut artifacts = vec![artifact_event];
-                    while let Ok(artifact) = receiver.try_recv() {
+                    while let StreamState::Value(artifact) =
+                        std::future::poll_fn(|cx| match inbound_stream.as_mut().poll_next(cx) {
+                            Poll::Pending => Poll::Ready(StreamState::NoNewValueAvailable),
+                            Poll::Ready(Some(artifact)) => {
+                                Poll::Ready(StreamState::Value(artifact))
+                            }
+                            Poll::Ready(None) => Poll::Ready(StreamState::EndOfStream),
+                        })
+                        .await
+                    {
```

Comment:
You can rewrite this to use [next()](https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.next) instead of manually implementing polling.

```suggestion
                    while let Some(artifact) = inbound_stream.next().await {
```

Reply:
No, the whole idea here is that we don't await at all and fetch whatever is already available. 

If you use a next(), then you will await on the future until something is available.

Reply:
One additional question. Why introduce async here with a block_on call if we are in a single threaded loop?